File: ruby/rotational-cipher/rotational_cipher.rb
# frozen_string_literal: true

module RotationalCipher
  ALPHABET_LENGTH = 26
  private_constant :ALPHABET_LENGTH
  ALPHA_CHARACTER = /[A-Za-z]/.freeze
  private_constant :ALPHA_CHARACTER
  DEFAULT_BASE = 0
  private_constant :DEFAULT_BASE
  INITIAL_CIPHER = ""
  private_constant :INITIAL_CIPHER
  IS_LOWERCASE = ->(char) { char.between?("a", "z") }
  private_constant :IS_LOWERCASE
  IS_UPPERCASE = ->(char) { char.between?("A", "Z") }
  private_constant :IS_UPPERCASE
  LOWER_BASE = "a".ord
  private_constant :LOWER_BASE
  UPPER_BASE = "A".ord
  private_constant :UPPER_BASE

  module_function

  def rotate(string, key)
    string.chars.reduce(INITIAL_CIPHER) do |acc, char|
      next (acc + char) unless char.match?(ALPHA_CHARACTER)

      acc + rotate_character(char, key)
    end
  end

  def rotate_character(char, key)
    base = base_for(char)
    char
      .ord
      .then { |num| num - base + key }
      .modulo(ALPHABET_LENGTH)
      .then { |num| num + base }
      .chr
  end
  private_class_method :rotate_character

  def base_for(char)
    case char
    when IS_LOWERCASE
      LOWER_BASE
    when IS_UPPERCASE
      UPPER_BASE
    else
      DEFAULT_BASE
    end
  end
  private_class_method :base_for
end

File: elixir/rotational-cipher/lib/rotational_cipher.ex
defmodule RotationalCipher do
  @alphabet_length 26

  defguardp is_lowercase?(char) when char in ?a..?z
  defguardp is_uppercase?(char) when char in ?A..?Z

  @doc """
  Given a plaintext and amount to shift by, return a rotated string.

  Example:
  iex> RotationalCipher.rotate("Attack at dawn", 13)
  "Nggnpx ng qnja"
  """
  @spec rotate(text :: String.t(), shift :: integer) :: String.t()
  def rotate(text, shift) do
    text
    |> String.to_charlist()
    |> Enum.reduce([], &rotate(shift, &1, &2))
    |> Enum.reverse()
    |> List.to_string()
  end

  defp rotate(shift, char, acc) when is_lowercase?(char) do
    [rotate_char(?a, char, shift) | acc]
  end

  defp rotate(shift, char, acc) when is_uppercase?(char) do
    [rotate_char(?A, char, shift) | acc]
  end

  defp rotate(_shift, char, acc), do: [char | acc]

  defp rotate_char(base, char, shift) do
    rem(char - base + shift, @alphabet_length) + base
  end
end
